### PR TITLE
New version

### PR DESCRIPTION
**Pull request by thibautjombart**
This implements major changes and improvements, including:

* removed legacy code from the 2 tabs era
* reorganise the app for clarity
* inputs are now categorised in separate tabs
* outputs are now categorised in separate tabs
* outputs for: LoS distribution, doubling time distribution, main results
* data can be uploaded as a spreadsheet
* past admissions are now accounted for in bed forecasts
* improved doc, now split into different tabs
* provide a `xlsx` template for data input with validated fields and a separate tab for explanations

I hope we can get this released ASAP. When making reviews please indicate comments you feel should be addressed before the release.

### CODE REVIEW

**Comment by pearsonca** [comment]
Looks suitable for release. A few notes:
 - can probably make the "Welcome" tab a dismissable modal. People aren't going to revisit that.
 - no handling of junk csvs at the moment? Not needed this instant, but should convert to some sort of warning ASAP. People are going to start putting garbage in there, and complain when it doesn't tell them they provided garbage
 - I have a funding statement now, so I'll add that in on acks.

File: app/R/plot_doubling_distribution.R [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#' @param title a string to be added to the resulting ggplot
+#' 
+#' @author Sam Clifford
+#' 
+
+plot_doubling_distribution <- function(x, title = NULL) {
+  dat <- data.frame(x = x)  
+  ggplot2::ggplot(data = dat,
+                  ggplot2::aes(x = x)) +
+    ggplot2::geom_histogram(fill = cmmid_color, col = "white") +
```
Comment (pearsonca) [inline app/R/plot_doubling_distribution.R]: maybe specify bins or binwidth here, to silence the warning.

File: app/R/run_model.R [hunk 2]
```diff
@@ -30,32 +30,75 @@
...
-#'
-#' x <- run_model(Sys.Date(),
-#'                n_start = 66,
-#'                doubling = c(4.5, 5.1, 6, 4.8),
-#'                duration = 14,
-#'                r_los = r_duration,
-#'                n_sim = 10)
-#' x
-#' plot(x)
+#' r_duration <- function(n = 1) rgeom(n, .2) + 2
```
Comment (pearsonca) [inline app/R/run_model.R:33]: delete this dead code

**Review by pearsonca — APPROVED** [review]

**Comment by mert0248** [comment]
Looks really good. One issue is that if the coefficient of variation for LOS is set to zero this throws and "Error:NA/NaN argument" on the "Length of Stay" distribution tab and subsequent errors on the main output tab.

Also on the "inputs" page if I click on the data template link this throws a 404 error for me. There also look possible a couple of typos on this page:

"Date of the reference: defaults to current day
Number of admissions on that date: number of new admissions on the data of reference; defaults to 1"

I think "Date of reference" be "Date of admission" and "data of reference" should be "date of admission:?

File: app/R/plot_doubling_distribution.R (hunk 1 above)
Comment (thibautjombart) [reply app/R/plot_doubling_distribution.R]: Yep doing it right now.

File: app/R/run_model.R (hunk 2 above)
Comment (thibautjombart) [reply app/R/run_model.R:33]: Actually it was meant to be an example, I just forgot to update the comments from ## to #'

Comment (thibautjombart) [reply app/R/run_model.R:33]: Fixed as of 119827237a7486b6fa9f2d1160387d637a82e2ea

File: app/R/plot_doubling_distribution.R (hunk 1 above)
Comment (thibautjombart) [reply app/R/plot_doubling_distribution.R]: Done as of 119827237a7486b6fa9f2d1160387d637a82e2ea

File: app/include/inputs.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+#### Minimum data: admissions on a single day
+
+The **minimum data** to use the app is a number of new admissions reported on a
+given day, used as starting point for the growth model and subsequent simulation
+of hospitalisations. If only this data is provided, the model will project bed
+occupancy for **new admissions only**, i.e. **not accounting for cases admitted
+before the starting point**.
+
+* **Date of the reference**: defaults to current day
```
Comment (mert0248) [inline app/include/inputs.md]: I think this should be "Date of admission"

File: app/include/inputs.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+* **Number of admissions on that date**: number of new admissions on the data of
```
Comment (mert0248) [inline app/include/inputs.md]: "data of reference" should be "date of admission"

**Comment by thibautjombart** [comment]
> Looks suitable for release. A few notes:
> 
> * can probably make the "Welcome" tab a dismissable modal. People aren't going to revisit that.

I agree! I don't know how to do that, but happy to see this PR-ed in a later release.

> * no handling of junk csvs at the moment? Not needed this instant, but should convert to some sort of warning ASAP. People are going to start putting garbage in there, and complain when it doesn't tell them they provided garbage

You're right. We provide an xlsx template, with ample explanations, and validated fields... maybe we should not allow csv at all? 

> * I have a funding statement now, so I'll add that in on acks.

Cool!

File: app/include/inputs.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+data, you need to upload a spreadsheet (we recommend `.xlsx` format), with 2
+columns:
+
+* **date**: dates at which new admissions have been recorded, using the
+  *yyy-mm-dd* format; for instance, 1982-02-04 is the 4th February 1982
+  
+* **n_admissions**: number of new admissions on these days
+
+To avoid data entry issues, we recommend using our 
+<a href="https://github.com/thibautjombart/covid19_bed_occupancy/blob/master/app/extra/data_model.xlsx?raw=true"> data template</a>.
```
Comment (mert0248) [inline app/include/inputs.md:34]: This link generates a 404 error for me

**Comment by thibautjombart** [comment]
> Looks really good. One issue is that if the coefficient of variation for LOS is set to zero this throws and "Error:NA/NaN argument" on the "Length of Stay" distribution tab and subsequent errors on the main output tab.
> 
> Also on the "inputs" page if I click on the data template link this throws a 404 error for me. There also look possible a couple of typos on this page:
> 
> "Date of the reference: defaults to current day
> Number of admissions on that date: number of new admissions on the data of reference; defaults to 1"
> 
> I think "Date of reference" be "Date of admission" and "data of reference" should be "date of admission:?

Thanks, should be all fixed now. For the template file, it was because the path was expecting to be on the master branch, fixed now too.

File: app/include/inputs.md (hunk 5 above)
Comment (thibautjombart) [reply app/include/inputs.md:34]: Can you try now?

**Comment by thibautjombart** [comment]
@pearsonca I have added clearer steps for data upload, including a link to the template, at f302acafb8216760528a05a308d4d39ee177f897

I have disabled the import of anything else than xslx and xls

**Comment by mert0248** [comment]
Not at all critical but maybe would be nice if the "Generate results" button could persist across the "Length of Stay Distribution" "Doubling time distribution" and "Main results" tabs so that you can always see where to press to run the simulation - and if not on the "Main Results" tab then switch focus to there?

Also is there any way to give an a visual indication that a calculation is in progress when it takes more than a second or two?

**Comment by thibautjombart** [comment]
> Not at all critical but maybe would be nice if the "Generate results" button could persist across the "Length of Stay Distribution" "Doubling time distribution" and "Main results" tabs so that you can always see where to press to run the simulation - and if not on the "Main Results" tab then switch focus to there?
> 
> Also is there any way to give an a visual indication that a calculation is in progress when it takes more than a second or two?

I was thinking that too but have no idea to handle that. Generally making tabs flash whenever there's some changes there (for LoS and doubling time too). Maybe raise a separate issue? It may be a really hard problem within the shiny framework...

**Review by erees — APPROVED** [review]
I think it looks great - don't have any issues or comments

**Comment by samclifford** [comment]
> I was thinking that too but have no idea to handle that. Generally making tabs flash whenever there's some changes there (for LoS and doubling time too). Maybe raise a separate issue? It may be a really hard problem within the shiny framework...

https://shiny.rstudio.com/articles/progress.html The issue will be that we're looping inside a function and so will have to embed the progress bar within _that_ and ensure that the `withProgress()` call wraps the call to the right part of the server.

**Comment by thibautjombart** [comment]
@mert0248 if happy can you approve the PR (review -> approve)?

**Comment by thibautjombart** [comment]
> > I was thinking that too but have no idea to handle that. Generally making tabs flash whenever there's some changes there (for LoS and doubling time too). Maybe raise a separate issue? It may be a really hard problem within the shiny framework...
> 
> https://shiny.rstudio.com/articles/progress.html The issue will be that we're looping inside a function and so will have to embed the progress bar within _that_ and ensure that the `withProgress()` call wraps the call to the right part of the server.

Cool. I agree it would be a nice addition but should not hold the current fix. Could you put in a separate issue?

**Review by mert0248 — APPROVED** [review]

**Comment by thibautjombart** [comment]
Note: the padding problem (not sure that's the right term) when resizing windows (see https://github.com/thibautjombart/covid19_bed_occupancy/issues/58) is not specific of this new version. We need to resolve it, but I would not hold the new release for it.

**Review by samclifford — CHANGES_REQUESTED** [review]
I get `Warning: Error in seq.int: 'from' must be a finite number` when uploading the data.xlsx file with an additional row added for 20 admissions on 2020-05-12. I've double checked that it's stored in date format.

See also the text issue in #58.

Reference on model description page is labelled (1) but references have no numbering. There's only one reference on the page, so it's obvious, but convert to author-date in text. Bullet point three says "see below" but this description is now in another tab.

**Comment by thibautjombart** [comment]
> I get `Warning: Error in seq.int: 'from' must be a finite number` when uploading the data.xlsx file with an additional row added for 20 admissions on 2020-05-12. I've double checked that it's stored in date format.
> 
> See also the text issue in #58.
> 
> Reference on model description page is labelled (1) but references have no numbering. There's only one reference on the page, so it's obvious, but convert to author-date in text. Bullet point three says "see below" but this description is now in another tab.

Could you share the guilty xlsx? I can't seem to reproduce the issue.

**Comment by samclifford** [comment]
> > I get `Warning: Error in seq.int: 'from' must be a finite number` when uploading the data.xlsx file with an additional row added for 20 admissions on 2020-05-12. I've double checked that it's stored in date format.
> > See also the text issue in #58.
> > Reference on model description page is labelled (1) but references have no numbering. There's only one reference on the page, so it's obvious, but convert to author-date in text. Bullet point three says "see below" but this description is now in another tab.
> 
> Could you share the guilty xlsx? I can't seem to reproduce the issue.

[data_model.xlsx](https://github.com/thibautjombart/covid19_bed_occupancy/files/4450432/data_model.xlsx)

Huh, when I open it in Preview.app the values are

| date | n_admissions |
| --- | --- |
| 2020-03-75 | 5 |
2020-03-76 | 14 |
2020-03-77 | 51 |
2020-03-78 | 37 |
2020-05-133 | 20 |

I built the file in OfficeLibre, though, so it might be that that's caused the issue.

**Comment by samclifford** [comment]
> I suspect that the above day values are the yday. 

I have made the date conversion more robust at be7e0e03c013b1ffe99298c592d538685b6ceb45 but unsure what the above comes from.

It works for me:
image

Also, I built the file with open-office too.

**Review by samclifford — APPROVED** [review]

**Pull request merged**